### PR TITLE
DOC-667 Azure Data Lake Gen2 output

### PR DESCRIPTION
**Pull request by asimms41**
## Description

Resolves [DOC-667](https://redpandadata.atlassian.net/browse/DOC-667)
Review deadline: 31st October

Cloud version is enabled in this PR: https://github.com/redpanda-data/cloud-docs/pull/101

## Page previews

[azure_data_lake_gen2 output](https://deploy-preview-78--redpanda-connect.netlify.app/redpanda-connect/components/outputs/azure_data_lake_gen2/)

## Checks

- [ ] New feature
- [x] Content gap
- [ ] Support Follow-up
- [ ] Small fix (typos, links, copyedits, etc)

[DOC-667]: https://redpandadata.atlassian.net/browse/DOC-667?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: modules/components/pages/outputs/azure_data_lake_gen2.adoc [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+:categories: ["Services","Azure"]
+
+
+// © 2024 Redpanda Data Inc.
+
+
+component_type_dropdown::[]
+
+
+Sends message parts as files to an https://learn.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-introduction[Azure Data Lake Gen2^] file system. Each file is uploaded with the file name specified in the `path` field.
```
Comment (asimms41) [inline modules/components/pages/outputs/azure_data_lake_gen2.adoc:15]: I've linked out to the Microsoft docs. But is this the best section to link to?

File: modules/components/pages/outputs/azure_data_lake_gen2.adoc [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+//tag::single-source[]
+:type: output
+:page-beta: true
+
+:categories: ["Services","Azure"]
+
+
+// © 2024 Redpanda Data Inc.
+
+
```
Comment (asimms41) [inline modules/components/pages/outputs/azure_data_lake_gen2.adoc:11]: In some cases, I've updated the wording of the field descriptions. Please check carefully.

File: modules/components/pages/outputs/azure_data_lake_gen2.adoc [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+If you configure multiple authentication methods, the `storage_connection_string` takes precedence.
+
+== Performance
+
+This output benefits from sending multiple messages in flight in parallel for improved performance. You can tune the max number of in flight messages (or message batches) with the field `max_in_flight`.
+
+== Fields
+
+=== `storage_account`
```
Comment (asimms41) [inline modules/components/pages/outputs/azure_data_lake_gen2.adoc:55]: Could we give examples to show the format of the value that is expected?

File: modules/components/pages/outputs/azure_data_lake_gen2.adoc [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+=== `filesystem`
+
+The name of the data lake storage file system you want to upload messages to. This field supports xref:configuration:interpolation.adoc#bloblang-queries[interpolation functions].
+
+*Type*: `string`
+
+```yml
+# Examples
+filesystem: messages-${!timestamp("2006")}
```
Comment (asimms41) [inline modules/components/pages/outputs/azure_data_lake_gen2.adoc:98]: Wouldn't you still need to include the name of the filesystem? I had a quick flick through the Microsoft docs but couldn't easily find where to get this/it's format.

File: modules/components/pages/outputs/azure_data_lake_gen2.adoc (hunk 1 above)
Comment (ooesili) [reply modules/components/pages/outputs/azure_data_lake_gen2.adoc:15]: That looks good!

File: modules/components/pages/outputs/azure_data_lake_gen2.adoc (hunk 4 above)
Comment (ooesili) [reply modules/components/pages/outputs/azure_data_lake_gen2.adoc:98]: The example would generate filesystem names like `messages-2024` which is a valid filesystem name.

The format of the field can be found in [the REST API documentation](https://learn.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/filesystem/create?view=rest-storageservices-datalakestoragegen2-2019-12-12) for the API call to create a filesystem.

> The filesystem identifier. The value must start and end with a letter or number and must contain only letters, numbers, and the dash (-) character. Consecutive dashes are not permitted. All letters must be lowercase. The value must have between 3 and 63 characters.

Similar to the `storage_account` and related fields, filesystem names have the same requirements as the [`container`](https://docs.redpanda.com/redpanda-connect/components/inputs/azure_blob_storage/#container) field in the `aws_blob_storage` plugin.

For a little context, the Azure Data Lake Storage Gen2 service is built upon the Blob Storage service, so there are direct analogues between the two services:
- `storage_account` -> `storage_account`
- `container` -> `filesystem`
- `path` -> `path`

File: modules/components/pages/outputs/azure_data_lake_gen2.adoc (hunk 3 above)
Comment (ooesili) [reply modules/components/pages/outputs/azure_data_lake_gen2.adoc:55]: The values are copied 1-to-1 from Azure, i.e. there isn't a specific format the user would have to synthesize the value into, so I don't know if examples would be that useful. For what it's worth, it is the exact same set of fields used by [the `azure_blob_storage` plugin](https://docs.redpanda.com/redpanda-connect/components/inputs/azure_blob_storage/).

File: modules/components/pages/outputs/azure_data_lake_gen2.adoc (hunk 2 above)
Comment (ooesili) [reply modules/components/pages/outputs/azure_data_lake_gen2.adoc:11]: All of the fields should be supported in cloud.

File: modules/components/pages/outputs/azure_data_lake_gen2.adoc [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+path: ${!meta("kafka_key")}.json
+path: ${!json("doc.namespace")}/${!json("doc.id")}.json
+```
+=== `max_in_flight`
+
+The maximum number of messages to have in flight at a given time. Increase this number to improve throughput.
+
+*Type*: `int`
+
+*Default*: `64`
```
Comment (kbatuigas) [inline modules/components/pages/outputs/azure_data_lake_gen2.adoc:121]: Is there a limit by any chance?

**Review by kbatuigas — APPROVED** [review]
One question, but lgtm otherwise

File: modules/components/pages/outputs/azure_data_lake_gen2.adoc (hunk 5 above)
Comment (asimms41) [reply modules/components/pages/outputs/azure_data_lake_gen2.adoc:121]: There isn't an explicit limit but there would be a point a which performance plateaus so I'll add that to the second line.

**Review by ooesili — APPROVED** [review]

**Pull request merged**